### PR TITLE
Improve serialization performance tests

### PR DESCRIPTION
**Pull request by khdang**
- Refactor all serialization performance tests to remove duplicated code. Adding more perf tests for all serializers is also more straightforward. Update namespace so the tests show up properly in reports.
- Add a flag in test project to enable running tests without project references, which helps us run against Desktop facades for comparison purpose
- Add performance tests for Json.Net for comparison purpose

cc: @SGuyGe @shmao @zhenlan

### CODE REVIEW

**Comment by khdang** [comment]
@dotnet-bot test this please

File: src/System.Runtime.Serialization.Json/tests/Performance/DcjsDeserializationTests.cs [hunk 1]
```diff
@@ -3,139 +3,35 @@
...
-                dictOfIntString[i] = i.ToString();
-            }
-
-            return dictOfIntString;
+            return PerformanceTestCommon.PerformanceMemberData(SerializerType.DataContractJsonSerializer);
         }
 
         [Benchmark]
-        public void DeserilaizeDictionary128()
+        [MemberData(nameof(SerializeMemberData))]
```
Comment (shmao) [inline src/System.Runtime.Serialization.Json/tests/Performance/DcjsDeserializationTests.cs]: C# 6 syntax may not build in K enlistment. That was the case for WCF.

Comment (stephentoub) [inline src/System.Runtime.Serialization.Json/tests/Performance/DcjsDeserializationTests.cs]: We're using it in thousands of places across corefx now.

File: src/System.Runtime.Serialization.Json/tests/Performance/DcjsPerformanceTest.cs [hunk 2]
```diff
@@ -0,0 +1,86 @@
+﻿using System.IO;
+using Newtonsoft.Json;
+
+namespace System.Runtime.Serialization.Json.Tests.Performance
+{
+    internal class DcjsSerializerFactory : SerializerFactory
+    {
+        private static DcjsSerializerFactory _instance = null;
```
Comment (shmao) [inline src/System.Runtime.Serialization.Json/tests/Performance/DcjsPerformanceTest.cs]: Rename it to s_instance ?

File: src/System.Xml.XmlSerializer/tests/Performance/XsPerformanceTest.cs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        }
+
+        public override IPerfTestSerializer GetSerializer(SerializerType serializerType)
+        {
+            switch (serializerType)
+            {
+                case SerializerType.XmlSerializer:
+                    return new XsSerializer();
+                default:
+                    throw new NotImplementedException();
```
Comment (shmao) [inline src/System.Xml.XmlSerializer/tests/Performance/XsPerformanceTest.cs]: This doesn't seem to be the right exception to throw from here. Maybe NotSupportedException?

File: src/System.Runtime.Serialization.Json/tests/DataContractJsonSerializer.cs [hunk 4]
```diff
@@ -15,7 +15,7 @@
 using System.Xml;
 using System.Xml.Linq;
 using Xunit;
-
+using Newtonsoft.Json;
```
Comment (SGuyGe) [inline src/System.Runtime.Serialization.Json/tests/DataContractJsonSerializer.cs]: Is this necessary?

File: src/System.Runtime.Serialization.Json/tests/Performance/DcjsPerformanceTest.cs [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        }
+    }
+
+    internal class DcjsSerializer : IPerfTestSerializer
+    {
+        private DataContractJsonSerializer _serializer;
+
+        public void Deserialize(Stream stream)
+        {
+            // Assumption: Deserialize() is always called after Init()
```
Comment (mconnew) [inline src/System.Runtime.Serialization.Json/tests/Performance/DcjsPerformanceTest.cs]: Add asserts to verify this.

File: src/System.Runtime.Serialization.Json/tests/Performance/DcjsPerformanceTest.cs [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+using Newtonsoft.Json;
+
+namespace System.Runtime.Serialization.Json.Tests.Performance
+{
+    internal class DcjsSerializerFactory : SerializerFactory
+    {
+        private static DcjsSerializerFactory _instance = null;
+        public static DcjsSerializerFactory GetInstance()
+        {
+            if (_instance == null)
```
Comment (mconnew) [inline src/System.Runtime.Serialization.Json/tests/Performance/DcjsPerformanceTest.cs]: Lazy initialization pattern is overkill. Set _instance in a static constructor which will only get run the first time the class type is referenced. Or simply instantiate the factory and set _instance on the field declaration line as the compiler will implicitly place this in the static constructor anyway.

File: src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+    }
+
+    public interface IPerfTestSerializer
+    {
+        void Init(object obj);
+        void Serialize(object obj, Stream s);
+        void Deserialize(Stream s);
+    }
+
+    public abstract class SerializerFactory
```
Comment (mconnew) [inline src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs]: Why use an interface for IPerfTestSerializer and an abstract class for SerializerFactory? I would make them both interfaces as abstract classes are for when you need some implementation to exist in a common base class. Whether you prefer interfaces of base abstract classes, you should be consistent.

File: src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+            for (int i = 0; i < obj.Length; ++i)
+            {
+                obj[i] = (byte)(i % 256);
+            }
+            return obj;
+        }
+
+        public static Dictionary<int, string> CreateDictionaryOfIntString(int count)
+        {
+            Dictionary<int, string> dictOfIntString = new Dictionary<int, string>();
```
Comment (mconnew) [inline src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs]: Pass the count to the Dictionary constructor. It defaults to a size of 3, and needs to grow (which requires new allocation and copying) each time the dictionary gets close to capacity. This could skew your results by requiring extra GC work during the serialization/deserialization.

File: src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+            yield return new object[] { 1, TestType.SimpleType, 15, serializerType };
+            yield return new object[] { 10000, TestType.ISerializable, -1, serializerType };
+            yield return new object[] { 10000, TestType.XmlElement, -1, serializerType };
+        }
+
+        #endregion
+
+        #region Methods to create object for serialization tests
+
+        public static SimpleTypeWihtMoreProperties CreateObject(int height, int parentId, int currentId, int collectionSize, int childListSize)
```
Comment (mconnew) [inline src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs]: Nit: Typo - Wiht -> With

File: src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+        {
+            int index = parentId * childListSize + (currentId + 1);
+            var obj = new SimpleTypeWihtMoreProperties()
+            {
+                IntProperty = index,
+                StringProperty = index + " string value",
+                EnumProperty = (MyEnum)(index % (Enum.GetNames(typeof(MyEnum)).Length)),
```
Comment (mconnew) [inline src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs]: Enum.GetNames is expensive.

File: src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+                CollectionProperty = new List<string>(),
```
Comment (mconnew) [inline src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs]: Pass collectionSize to List constructor otherwise the list will reallocate to double it's size every time you add more than it's capacity. Default capacity is 2.

File: src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+                SimpleTypeList = new List<SimpleTypeWihtMoreProperties>()
```
Comment (mconnew) [inline src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs]: Pass childListSize to List constructor

File: src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+    public interface IPerfTestSerializer
+    {
+        void Init(object obj);
+        void Serialize(object obj, Stream s);
+        void Deserialize(Stream s);
+    }
+
+    public abstract class SerializerFactory
+    {
+        public abstract IPerfTestSerializer GetSerializer(SerializerType serializerType);
```
Comment (shmao) [inline src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs]: I wondered if it is necessary for GetSerializer to take serializerType as a parameter. We have three concrete factories. Two of them return only one type of IPerfTestSerializer. Only DcjsSerializerFactory may return two types of IPerfTestSerializer depending on the serializerType. Can we just have four concrete factories instead?

File: src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+            }
+            return obj;
+        }
+
+        public static byte[] CreateByteArray(int size)
+        {
+            byte[] obj = new byte[size];
+            for (int i = 0; i < obj.Length; ++i)
+            {
+                obj[i] = (byte)(i % 256);
```
Comment (mconnew) [inline src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs]: No need to do modulo division, cast to byte and mark as unchecked.

File: src/System.Runtime.Serialization.Json/tests/System.Runtime.Serialization.Json.Tests.csproj [hunk 15]
```diff
@@ -10,29 +10,36 @@
...
+    </ProjectReference>		
+    <ProjectReference Include="..\..\System.Runtime.Serialization.Primitives\src\System.Runtime.Serialization.Primitives.csproj">		
+      <Project>{CDF0ACB5-1361-4E48-8ECB-22E8022F5F01}</Project>		
+      <Name>System.Runtime.Serialization.Primitives</Name>		
+    </ProjectReference>		
+    <ProjectReference Include="..\..\System.Xml.XmlSerializer\src\System.Xml.XmlSerializer.csproj">		
+      <Project>{D62A6082-5229-4845-8BE9-75753E08C65A}</Project>		
+      <Name>System.Xml.XmlSerializer</Name>		
+    </ProjectReference>		
   </ItemGroup>
```
Comment (SGuyGe) [inline src/System.Runtime.Serialization.Json/tests/System.Runtime.Serialization.Json.Tests.csproj]: Nit: Remove trailing tabs.

File: src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+                    obj = new string('k', testSize);
+                    break;
+                case TestType.XmlElement:
+                    XmlDocument xmlDoc = new XmlDocument();
+                    xmlDoc.LoadXml(@"<html></html>");
+                    XmlElement xmlElement = xmlDoc.CreateElement("Element");
+                    xmlElement.InnerText = "Element innertext";
+                    obj = xmlElement;
+                    break;
+            }
```
Comment (shmao) [inline src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs]: In the future, we may add new test type. If we forgot to add a case entry for the new test type, this method would return null. Do we want to add an Assert or throw exception in default if the TestType is not recognized?

File: src/System.Runtime.Serialization.Json/tests/Performance/DcjsDeserializationTests.cs [hunk 17]
```diff
@@ -3,139 +3,35 @@
...
+        public void DcjsDeSerializationTest(int iterations, TestType testType, int testSize, SerializerType serializerType)
```
Comment (SGuyGe) [inline src/System.Runtime.Serialization.Json/tests/Performance/DcjsDeserializationTests.cs]: DeSerialization => Deserialization

File: src/System.Runtime.Serialization.Json/tests/Performance/DcjsDeserializationTests.cs [hunk 18]
```diff
@@ -3,139 +3,35 @@
...
         {
-            SimpleType simpleType = new SimpleType() { P1 = "Foo", P2 = 123 };
-            RunDcjsDeserializationTest(simpleType, 10000);
+            return PerformanceTestCommon.PerformanceMemberData(SerializerType.JsonNet);
         }
 
         [Benchmark]
-        public void DeserializeIXmlSerializable()
+        [MemberData(nameof(SerializeMemberDataJsonNet))]
+        public void JsonNetDeSerializationTest(int iterations, TestType testType, int testSize, SerializerType serializerType)
```
Comment (SGuyGe) [inline src/System.Runtime.Serialization.Json/tests/Performance/DcjsDeserializationTests.cs]: DeSerialization => Deserialization

File: src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs [hunk 19]
```diff
@@ -0,0 +1,207 @@
...
+        #region Methods to run serialization performance tests
+
+        public static void RunSerializationPerformanceTest(int iterations, SerializerType serializerType, TestType testType, int testSize, SerializerFactory serializerFactory)
+        {
+            var obj = GetSerializationObject(testType, testSize);
+
+            var serializer = serializerFactory.GetSerializer(serializerType);
+            serializer.Init(obj);
+
+            using (var stream = new MemoryStream())
```
Comment (mconnew) [inline src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs]: MemoryStream starts with a size of 0, then allocates 256 bytes on the first write. It then doubles in size and needs to copy the contents each time you exceed the capacity. As you are continuously writing to the same stream without resetting it with each iteration, you are causing the stream to hold all results of all tests. The ideal thing would be to know the maximum size stream you would need and preallocate a byte array before the test starts and pass to the MemoryStream constructor. This way MemoryStream will throw an exception if you try to exceed that and you guarantee there's no allocation coming from MemoryStream. An alternative solution would be to do a priming serialization after the MemoryStream has been allocated so that MemoryStream has grown sufficiently to avoid allocations before measurements begin. You could still end up with GC from any allocations potentially happening during the test, but it would be a single GC. Regardless which option you use, you should reset the stream before each serialization with `stream.Position = 0;`.

File: src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs [hunk 20]
```diff
@@ -0,0 +1,207 @@
...
+                        for (int i = 0; i < iterations; i++)
+                        {
+                            serializer.Serialize(obj, stream);
+                        }
+                    }
+                }
+            }
+        }
+
+        public static void RunDeSerializationPerformanceTest(int iterations, SerializerType serializerType, TestType testType, int testSize, SerializerFactory serializerFactory)
```
Comment (SGuyGe) [inline src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs]: DeSerialization => Deserialization

File: src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs [hunk 21]
```diff
@@ -0,0 +1,207 @@
...
+            var serializer = serializerFactory.GetSerializer(serializerType);
+            serializer.Init(obj);
+
+            using (var stream = new MemoryStream())
+            {
+                foreach (var iteration in Benchmark.Iterations)
+                {
+                    using (iteration.StartMeasurement())
+                    {
+                        for (int i = 0; i < iterations; i++)
```
Comment (mconnew) [inline src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs]: The word iteration is being overloaded. Benchmark.Iterations is referring to an iteration of running the Benchmark, whereas the iterations here is referring to the count of the number of times your are serializing. It's a little confusing.

File: src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs [hunk 22]
```diff
@@ -0,0 +1,207 @@
...
+        }
+
+        public static void RunDeSerializationPerformanceTest(int iterations, SerializerType serializerType, TestType testType, int testSize, SerializerFactory serializerFactory)
+        {
+            var obj = GetSerializationObject(testType, testSize);
+
+            var serializer = serializerFactory.GetSerializer(serializerType);
+            serializer.Init(obj);
+
+            using (var stream = new MemoryStream())
```
Comment (mconnew) [inline src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs]: Same issues with the MemoryStream here where it's just growing and growing.

File: src/System.Xml.XmlSerializer/tests/Performance/XsPerformanceTest.cs (hunk 3 above)
Comment (mconnew) [inline src/System.Xml.XmlSerializer/tests/Performance/XsPerformanceTest.cs]: I don't think either of them are quite appropriate as you are saying either the method is not implemented of it's not supported which isn't the case. I think either ArgumentException or ArgumentOutOfRangeException would be better as the problem is with what was requested and not that the GetSerializer method isn't functional. Either way though, this is test code so it's not super important :smile:

File: src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs (hunk 9 above)
Comment (SGuyGe) [inline src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs]: The name is too generic. => CreateSimpleTypeWihtMoreProperties

File: src/System.Runtime.Serialization.Json/tests/Performance/DcjsPerformanceTest.cs [hunk 23]
```diff
@@ -0,0 +1,86 @@
...
+        public void Init(object obj)
+        {
+            _serializer = new JsonSerializer();
+        }
+
+        public void Serialize(object obj, Stream stream)
+        {
+            // Assumption: Serialize() is always called after Init()
+            // Assumption: stream != null and stream position will be reset to 0 after this method
+            var writer = new StreamWriter(stream);
```
Comment (mconnew) [inline src/System.Runtime.Serialization.Json/tests/Performance/DcjsPerformanceTest.cs]: Use StreamWriter in a using block and the Dispose will do the Flush for you.

File: src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs [hunk 24]
```diff
@@ -0,0 +1,207 @@
...
+
+            return dictOfIntString;
+        }
+
+        public static List<int> CreateListOfInt(int count)
+        {
+            return Enumerable.Range(0, count).ToList();
+        }
+
+        public static object GetSerializationObject(TestType testType, int testSize)
```
Comment (SGuyGe) [inline src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs]: I don't see much value of having TestType, it would only add extra maintenance cost. I think we should encapsulate all information related to a certain test into a separate type, that way we can simply adding new types for new tests.

File: src/System.Runtime.Serialization.Json/tests/Performance/DcjsPerformanceTest.cs (hunk 23 above)
Comment (shmao) [inline src/System.Runtime.Serialization.Json/tests/Performance/DcjsPerformanceTest.cs]: Dispose would also close the underlying stream, right? I think we don't want to close the stream here as we may use it in Deserialize() later.

File: src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs (hunk 24 above)
Comment (khdang) [inline src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs]: It's for the representation of the result. Using TestType enum would show up nicely in the output xml file. It seems that using object type as a parameter for MemberData would result in all tests are being ignored, apparently due to errors. Let me try overriding the toString() method.

File: src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs [hunk 25]
```diff
@@ -0,0 +1,207 @@
...
+
+    public class PerformanceTestCommon
+    {
+
+        #region Performance test configuration
+
+        public static IEnumerable<object[]> PerformanceMemberData(SerializerType serializerType)
+        {
+            yield return new object[] { 100, TestType.ByteArray, 1024, serializerType };
+            yield return new object[] { 5, TestType.ByteArray, 1024 * 1024, serializerType };
```
Comment (SGuyGe) [inline src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs]: 1. It's hard to tell what each element is for. 2. Can we eliminate serializerType as it's always the same? 3. We should encapsulate the object creation logic along with other test settings (not necessarily here) and eliminate TestType.

File: src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs (hunk 21 above)
Comment (khdang) [inline src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs]: Agree. I will rename it.
will rename it.

File: src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs (hunk 25 above)
Comment (khdang) [inline src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs]: 1. Each line is a configuration for a scenario.
2. This method would take the serializerType as parameter and we have 4 different serializer types so it needs to be there.
3. Agree, that would be nice. But for me, there was an issue when I did that. Object type when passed into MemberData causes some strange issue that the tests don't get run.

File: src/System.Runtime.Serialization.Xml/tests/System.Runtime.Serialization.Xml.Tests.csproj [hunk 26]
```diff
@@ -27,24 +34,23 @@
     <Compile Include="XmlDictionaryReaderTests.cs" />
     <Compile Include="XmlDictionaryWriterTest.cs" />
     <!-- Performance Tests -->
-    <Compile Include="Performance\Perf.Deserialization.cs" />
     <Compile Include="$(CommonTestPath)\System\PerfUtils.cs">
```
Comment (SGuyGe) [inline src/System.Runtime.Serialization.Xml/tests/System.Runtime.Serialization.Xml.Tests.csproj]: Why deleting this?

File: src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs (hunk 19 above)
Comment (khdang) [inline src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs]: Yes you're right the stream position reset should be here instead of in the serializer.

File: src/System.Runtime.Serialization.Xml/tests/System.Runtime.Serialization.Xml.Tests.csproj (hunk 26 above)
Comment (khdang) [inline src/System.Runtime.Serialization.Xml/tests/System.Runtime.Serialization.Xml.Tests.csproj]: This has been refactored into this configuration :): 
```
yield return new object[] { 1, TestType.SimpleType, 15, serializerType };
```

File: src/System.Xml.XmlSerializer/tests/Performance/XsPerformanceTest.cs [hunk 27]
```diff
@@ -0,0 +1,55 @@
...
+            }
+        }
```
Comment (SGuyGe) [inline src/System.Xml.XmlSerializer/tests/Performance/XsPerformanceTest.cs]: It doesn't make sense to pass in a parameter while the code only expects a specific value. Create separate factories for DCJS and JsonNet and eliminate SerializerType.

File: src/System.Runtime.Serialization.Json/tests/Performance/DcjsPerformanceTest.cs (hunk 2 above)
Comment (khdang) [inline src/System.Runtime.Serialization.Json/tests/Performance/DcjsPerformanceTest.cs]: Yup

File: src/System.Xml.XmlSerializer/tests/System.Xml.XmlSerializer.Tests.csproj [hunk 28]
```diff
@@ -10,30 +10,37 @@
     <RootNamespace>System.Xml.XmlSerializer.Tests</RootNamespace>
     <AssemblyName>System.Xml.XmlSerializer.Tests</AssemblyName>
     <ProjectGuid>{4050F1D1-1DD2-4B48-A17B-E3F90DD18C4B}</ProjectGuid>
+    <UseProjectReferences>true</UseProjectReferences>
+  </PropertyGroup>
+  <PropertyGroup Condition="!$(UseProjectReferences)" >
+    <ProjectJson>Performance/project.json</ProjectJson>
+    <ProjectLockJson>Performance/project.lock.json</ProjectLockJson>
```
Comment (SGuyGe) [inline src/System.Xml.XmlSerializer/tests/System.Xml.XmlSerializer.Tests.csproj]: Why do we need 2 project.json?

File: src/System.Runtime.Serialization.Xml/tests/System.Runtime.Serialization.Xml.Tests.csproj [hunk 29]
```diff
@@ -11,11 +11,18 @@
     <AssemblyName>System.Runtime.Serialization.Xml.Tests</AssemblyName>
     <ProjectGuid>{30CAB353-089E-4294-B23B-F2DD1D945654}</ProjectGuid>
     <IncludePerformanceTests>true</IncludePerformanceTests>
+    <UseProjectReferences>true</UseProjectReferences>
+  </PropertyGroup>
+  <PropertyGroup Condition="!$(UseProjectReferences)" >
+    <ProjectJson>Performance/project.json</ProjectJson>
```
Comment (SGuyGe) [inline src/System.Runtime.Serialization.Xml/tests/System.Runtime.Serialization.Xml.Tests.csproj]: Why do we need 2 project.json?

File: src/System.Runtime.Serialization.Json/tests/System.Runtime.Serialization.Json.Tests.csproj [hunk 30]
```diff
@@ -10,29 +10,36 @@
     <RootNamespace>System.Runtime.Serialization.Json.Tests</RootNamespace>
     <AssemblyName>System.Runtime.Serialization.Json.Tests</AssemblyName>
     <ProjectGuid>{701CB3BC-00DC-435D-BDE4-C5FC29A708A7}</ProjectGuid>
+    <UseProjectReferences>true</UseProjectReferences>
+  </PropertyGroup>
+  <PropertyGroup Condition="!$(UseProjectReferences)" >
+    <ProjectJson>Performance/project.json</ProjectJson>
```
Comment (SGuyGe) [inline src/System.Runtime.Serialization.Json/tests/System.Runtime.Serialization.Json.Tests.csproj]: Why do we need 2 project.json?

File: src/System.Xml.XmlSerializer/tests/Performance/XsPerformanceTest.cs (hunk 27 above)
Comment (khdang) [inline src/System.Xml.XmlSerializer/tests/Performance/XsPerformanceTest.cs]: I was thinking of having all performance tests in one project but decide not to do that. So yes we should remove the serializerType now. Shin also has the same suggestion.

File: src/System.Runtime.Serialization.Xml/tests/System.Runtime.Serialization.Xml.Tests.csproj [hunk 31]
```diff
@@ -11,11 +11,18 @@
...
+    <ProjectLockJson>Performance/project.lock.json</ProjectLockJson>
```
Comment (shmao) [inline src/System.Runtime.Serialization.Xml/tests/System.Runtime.Serialization.Xml.Tests.csproj]: Do we need to reference project.lock.json?

File: src/System.Runtime.Serialization.Json/tests/Performance/DcjsPerformanceTest.cs (hunk 23 above)
Comment (mconnew) [inline src/System.Runtime.Serialization.Json/tests/Performance/DcjsPerformanceTest.cs]: StreamWriter has an overload to specify whether to leave the stream open or not when you dispose it.

File: src/System.Runtime.Serialization.Json/tests/Performance/DcjsDeserializationTests.cs [hunk 32]
```diff
@@ -3,139 +3,35 @@
...
-            for (int i = 0; i < count; ++i)
-            {
-                dictOfIntString[i] = i.ToString();
-            }
-
-            return dictOfIntString;
+            return PerformanceTestCommon.PerformanceMemberData(SerializerType.DataContractJsonSerializer);
         }
 
         [Benchmark]
```
Comment (brianrob) [inline src/System.Runtime.Serialization.Json/tests/Performance/DcjsDeserializationTests.cs]: I would recommend looking at an example performance test in corefx (e.g. src\System.Collections\tests\Performance) and the README for xunit.performance at [https://github.com/Microsoft/xunit-performance/blob/master/README.md.](https://github.com/Microsoft/xunit-performance/blob/master/README.md.)

In addition to using the [Benchmark] attribute, you will also need to either use Benchmark.Iterate() or create an outer loop to allow the performance harness to execute multiple iterations.  If you choose to create a loop, you'll also need to call StartMeasurement() to start measuring each iteration after any required setup.

The README for xunit.performance has an example of a simple performance test which illustrates exactly what needs to be done to integrate with the performance harness.

Comment (khdang) [inline src/System.Runtime.Serialization.Json/tests/Performance/DcjsDeserializationTests.cs]: @brianrob Yes I've followed the guideline for perf tests and we do use Benchmark.Ierations and StartMeasurement in the main method to run serialization perf test at: https://github.com/dotnet/corefx/pull/7119/files#diff-392ee56ad323b27581f82a43af91af89R160

File: src/System.Runtime.Serialization.Json/tests/System.Runtime.Serialization.Json.Tests.csproj (hunk 30 above)
Comment (khdang) [inline src/System.Runtime.Serialization.Json/tests/System.Runtime.Serialization.Json.Tests.csproj]: We need project.json and project.lock.json.

File: src/System.Runtime.Serialization.Json/tests/DataContractJsonSerializer.cs (hunk 4 above)
Comment (khdang) [inline src/System.Runtime.Serialization.Json/tests/DataContractJsonSerializer.cs]: No I've removed it.

File: src/System.Runtime.Serialization.Json/tests/System.Runtime.Serialization.Json.Tests.csproj (hunk 15 above)
Comment (khdang) [inline src/System.Runtime.Serialization.Json/tests/System.Runtime.Serialization.Json.Tests.csproj]: Done. Removed trailing tabs, in other csproj files as well

File: src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs (hunk 13 above)
Comment (khdang) [inline src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs]: That's true. I've remove serializerType so now each factory is in charge of creating only one serializer type.

File: src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs (hunk 7 above)
Comment (khdang) [inline src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs]: I don't have a good reason. I've updated this to interface as well.

File: src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs (hunk 24 above)
Comment (khdang) [inline src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs]: I tried refactor the test into a type but there is a problem with MemberData attribute. We can't seem to use object for Theory parameter in performance test because the test will go missing from the result. I'll just keep this part of the code as it is now.

File: src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs (hunk 16 above)
Comment (khdang) [inline src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs]: I've added exception if the test type is not recognized.

File: src/System.Runtime.Serialization.Xml/tests/System.Runtime.Serialization.Xml.Tests.csproj (hunk 31 above)
Comment (khdang) [inline src/System.Runtime.Serialization.Xml/tests/System.Runtime.Serialization.Xml.Tests.csproj]: Yes we need reference to project.lock.json.

File: src/System.Xml.XmlSerializer/tests/Performance/XsPerformanceTest.cs (hunk 3 above)
Comment (khdang) [inline src/System.Xml.XmlSerializer/tests/Performance/XsPerformanceTest.cs]: I've refactored this part and now each factory creates only one type of serializer.

File: src/System.Xml.XmlSerializer/tests/System.Xml.XmlSerializer.Tests.csproj (hunk 28 above)
Comment (khdang) [inline src/System.Xml.XmlSerializer/tests/System.Xml.XmlSerializer.Tests.csproj]: We need project.json and project.lock.json.

File: src/System.Runtime.Serialization.Xml/tests/System.Runtime.Serialization.Xml.Tests.csproj (hunk 29 above)
Comment (khdang) [inline src/System.Runtime.Serialization.Xml/tests/System.Runtime.Serialization.Xml.Tests.csproj]: We need project.json and project.lock.json.

**Comment by khdang** [comment]
I've committed some changes to address the comments.

File: src/System.Runtime.Serialization.Json/tests/System.Runtime.Serialization.Json.Tests.csproj (hunk 30 above)
Comment (khdang) [inline src/System.Runtime.Serialization.Json/tests/System.Runtime.Serialization.Json.Tests.csproj]: To clarify regarding project.lock.json, we need to have the project.lock.json entry in the csproj in order build successfully. Otherwise the original project.lock.json is used and errors will be thrown due to types not found. We're not checking in any project.lock.json files.

**Comment by khdang** [comment]
@dotnet-bot test this please

**Comment by khdang** [comment]
I'm not sure why the build validations are not running so I'm closing and will reopen this PR.

File: src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs (hunk 10 above)
Comment (khdang) [inline src/System.Runtime.Serialization.Xml/tests/Performance/PerformanceTestsCommon.cs]: Thanks for the insight. For now this only runs once so it won't be an issue. I will keep a note and make adjustment when we have more tests that run this part more often.

File: src/System.Xml.XmlSerializer/tests/System.Xml.XmlSerializer.Tests.csproj (hunk 28 above)
Comment (khdang) [inline src/System.Xml.XmlSerializer/tests/System.Xml.XmlSerializer.Tests.csproj]: As the above comment, we need an entry of project.lock.json and we're not checking in any project.lock.json file.
